pkg/util/kubernetes/leaderelection/leaderelection.go

```diff
@@ -0,0 +1,216 @@
+	globalLeaderEngine = nil
+}
+
+// SetLeaderLeaseDuration is a helper to set the current LeaderLeaseDuration global
+// It is ONLY to be used for tests
+func SetLeaderLeaseDuration(ttl time.Duration) {
+	globalLeaderLeaseDuration = ttl
+}
+
+// SetHolderIdentify is a helper to set the current holderIdentify global
+// It is ONLY to be used for tests
+func SetHolderIdentify(holderIdentity string) {
+	globalHolderIdentity = holderIdentity
+}
+
+func GetLeaderEngine() (*LeaderEngine, error) {
+	if globalLeaderEngine == nil {
+		globalLeaderEngine = newLeaderEngine()
+		globalLeaderEngine.initRetry.SetupRetrier(&retry.Config{
+			Name:          "leaderElection",
+			AttemptMethod: globalLeaderEngine.init,
+			Strategy:      retry.RetryCount,
+			RetryCount:    10,
+			RetryDelay:    30 * time.Second,
+		})
+	}
+	err := globalLeaderEngine.initRetry.TriggerRetry()
+	if err != nil {
+		log.Debugf("Init error: %s", err)
+		return nil, err
+	}
+	return globalLeaderEngine, nil
+}
+
+func (le *LeaderEngine) init() error {
+	var err error
+
+	if globalHolderIdentity == "" {
+		globalHolderIdentity, err = os.Hostname()
+		if err != nil {
+			log.Debugf("cannot get hostname: %s", err)
+			return err
+		}
+	}
+	le.HolderIdentity = globalHolderIdentity
+	log.Debugf("HolderIdentity is %q", globalHolderIdentity)
+
+	leaseDuration := config.Datadog.GetInt("leader_lease_duration")
+	if leaseDuration != 0 {
+		globalLeaderLeaseDuration = time.Duration(leaseDuration) * time.Second
+	}
+
+	if globalLeaderLeaseDuration == 0 {
+		globalLeaderLeaseDuration = defaultLeaderLeaseDuration
+	}
+	le.LeaseDuration = globalLeaderLeaseDuration
+	log.Debugf("LeaderLeaseDuration is %s", globalLeaderLeaseDuration.String())
+
+	le.coreClient, err = GetClient()
+	if err != nil {
+		log.Errorf("Not Able to set up a client for the Leader Election: %s", err)
+		return err
+	}
+
+	// check if we can get endpoints.
+	_, err = le.coreClient.Endpoints(metav1.NamespaceDefault).List(metav1.ListOptions{Limit: 1})
+	if err != nil {
+		log.Errorf("Cannot retrieve endpoints from the %s namespace", metav1.NamespaceDefault)
+		return err
+	}
+
+	le.leaderElector, err = le.newElection(le.LeaseName, metav1.NamespaceDefault, le.LeaseDuration)
+	if err != nil {
+		log.Errorf("Could not initialize the Leader Election process: %s", err.Error())
+		return err
+	}
+	log.Debug("Kubernetes official client successfully initialized")
+	return nil
+}
+
+// RunElection runs an election given an leader elector. Doesn't return.
+// The passed LeaderElector embeds callback functions that are triggered to handle the different states of the process.
+func (le *LeaderEngine) startLeaderElection() {
+	log.Infof("Starting Leader Election process for %q ...", le.HolderIdentity)
+	go wait.Forever(le.leaderElector.Run, 0)
+}
+
+// EnsureLeaderElectionRuns
+func (le *LeaderEngine) EnsureLeaderElectionRuns() error {
+	var leaderIdentity string
+
+	le.once.Do(le.startLeaderElection)
+	timeout := time.After(time.Second * 5)
```

Comment:
Of course it can be a connectivity issue/timeout issue in the client, but also something else.

I can set the timeout of this waiting block to the client timeout * 2.